bundles/org.openhab.binding.magentatv/src/main/java/org/openhab/binding/magentatv/internal/network/MagentaTVOAuth.java

```diff
@@ -0,0 +1,200 @@
+import java.util.UUID;
+
+import javax.ws.rs.HttpMethod;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.io.net.http.HttpUtil;
+import org.openhab.binding.magentatv.internal.MagentaTVException;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OAuthAuthenticateResponse;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OAuthAuthenticateResponseInstanceCreator;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OAuthTokenResponse;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OAuthTokenResponseInstanceCreator;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OauthCredentials;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OauthCredentialsInstanceCreator;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OauthKeyValue;
+import org.openhab.binding.magentatv.internal.handler.MagentaTVControl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * The {@link MagentaTVOAuth} class implements the OAuth authentication, which
+ * is used to query the userID from the Telekom platform.
+ *
+ * @author Markus Michels - Initial contribution
+ *
+ *         Deutsche Telekom uses a OAuth-based authentication to access the EPG portal. The
+ *         communication between the MR and the remote app requires a pairing before the receiver could be
+ *         controlled by sending keys etc. The so called userID is not directly derived from any local parameters
+ *         (like terminalID as a has from the mac address), but will be returned as a result from the OAuth
+ *         authentication. This will be performed in 3 steps
+ *         1. Get OAuth credentials -> Service URL, Scope, Secret, Client ID
+ *         2. Get OAth Token -> authentication token for step 3
+ *         3. Authenticate, which then provides the userID (beside other parameters)
+ *
+ */
+@NonNullByDefault
+public class MagentaTVOAuth {
+    private final Logger logger = LoggerFactory.getLogger(MagentaTVOAuth.class);
+
+    public String getUserId(String accountName, String accountPassword) throws MagentaTVException {
+        logger.debug("Authenticate with account {}", accountName);
+        if (accountName.isEmpty() || accountPassword.isEmpty()) {
+            throw new MagentaTVException("Credentials for OAuth missing, check thing config!");
+        }
+
+        String step = "initialize";
+        String url = "";
+        Properties httpHeader;
+        String postData = "";
+        String httpResponse = "";
+        InputStream dataStream = null;
+
+        // OAuth autentication results
+        String oAuthScope = "";
+        String oAuthService = "";
+        String epghttpsurl = "";
+        String retcode = "";
+        String retmsg = "";
+
+        try {
+            Gson gson = new GsonBuilder()
```

Comment:
The gson variable and initialization can also be a field, so it's only done once.

Reply:
changed